src/software/geom/algorithms/step_along_perimeter.cpp

```diff
@@ -0,0 +1,76 @@
+#include "software/geom/algorithms/step_along_perimeter.h"
+
+#include <numeric>
+#include <vector>
+
+#include "software/geom/algorithms/collinear.h"
+#include "software/geom/algorithms/distance.h"
+#include "software/geom/segment.h"
+
+
+Point stepAlongPerimeter(const Polygon& polygon, const Point& start,
+                         double travelDistance)
+{
+    if (travelDistance == 0.0)
+    {
+        return start;
+    }
+
+    std::vector<Segment> polygonSegments = polygon.getSegments();
+    std::size_t startSegmentIdx          = 0;
+
+    int i = 0;
+
+    // find initial segment which contains start point
+    for (Segment& segment : polygonSegments)
+    {
+        Point segmentStart = segment.getStart();
+        Point segmentEnd   = segment.getEnd();
+
+        if (collinear(segmentStart, start, segmentEnd))
+        {
+            startSegmentIdx = i;
+            break;
+        }
+        i++;
+    }
```

Comment:
Don't have to do this, but we could use [`std::find_if`](https://en.cppreference.com/w/cpp/algorithm/find) here

Reply:
ill rewrite this using find_if but why do you say we would not have to do this? is this step redundant?

Reply:
It's basically just syntactic sugar, so it's up to personal preference whether you think find_if looks better or not

Reply:
oh yeah okay. i replaced it with find_if, i thought you meant that the process of finding the segment containing the start is redundant.